controller/server/main.go

```diff
@@ -183,35 +192,69 @@ func validatePath(path string) (string, error) {
 
 func (s *server) CreateTopology(ctx context.Context, req *cpb.CreateTopologyRequest) (*cpb.CreateTopologyResponse, error) {
 	log.Infof("Received CreateTopology request: %v", req)
-	bp, err := validatePath(req.Topology.Name)
-	if err != nil {
-		return nil, status.Errorf(codes.InvalidArgument, "invalid topology path: %v", err)
+	topoPb := req.GetTopology()
+	if topoPb == nil {
+		return nil, status.Errorf(codes.InvalidArgument, "invalid request: missing topology protobuf")
 	}
-	p := topo.TopologyParams{
-		TopoName:       req.Topology.Name,
-		TopoNewOptions: []topo.Option{topo.WithBasePath(filepath.Dir(bp))},
-		Kubecfg:        req.Kubecfg,
+	if topoPb.GetName() == "" {
+		return nil, status.Errorf(codes.InvalidArgument, "missing topology name")
+	}
+
+	s.muTopo.Lock()
+	defer s.muTopo.Unlock()
+	if _, ok := s.topos[topoPb.GetName()]; ok {
+		return nil, status.Errorf(codes.AlreadyExists, "topology %q already exists", req.Topology.GetName())
+	}
+
+	log.Infof("Validating the topology protobuf...")
+	for _, node := range topoPb.Nodes {
+		if node.GetConfig() == nil || node.GetConfig().GetFile() == "" {
+			// A config section is not required: you are allowed to bring up a
+			// topology with no initial config.
+			continue
+		}
+		log.Infof("Check config path: %q", node.GetConfig().GetFile())
+		if _, err := validatePath(node.GetConfig().GetFile()); err != nil {
+			return nil, status.Errorf(codes.NotFound, "config file not found for node %q: %v", node.GetName(), err)
+		}
+	}
+	// Saves the original topology protobuf.
+	txtPb, err := prototext.Marshal(topoPb)
+	if err != nil {
+		return nil, status.Errorf(codes.Internal, "invalid topology protobuf: %v", err)
 	}
 
-	if err = topo.CreateTopology(ctx, p); err != nil {
+	if err := topo.CreateTopology(ctx, topo.TopologyParams{
+		TopoNewOptions: []topo.Option{topo.WithTopology(topoPb)},
+		Kubecfg:        req.Kubecfg,
```

Comment:
this path needs to validated and use the default value if not valid

Reply:
Done.